### ISSUE TITLE
MPS not working properly

### ISSUE THREAD

**Issue by Jerry-Master** [nerfstudio-project/nerfstudio#2435]
**Describe the bug**
The training does not work on M1 Macbook Pro.

**To Reproduce**
1. Process data: ns-process-data images --data trial1/images/ --output-dir trial1/output/ --num-downscales 0 --no-gpu --matching-method exhaustive
2. Run training: ns-train nerfacto --data trial1/output/  --machine.device-type mps --mixed-precision False

**Expected behavior**
It gives this error:

```
Traceback (most recent call last):
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/bin/ns-train", line 10, in <module>
    sys.exit(entrypoint())
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/scripts/train.py", line 261, in entrypoint
    main(
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/scripts/train.py", line 246, in main
    launch(
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/scripts/train.py", line 189, in launch
    main_func(local_rank=0, world_size=world_size, config=config)
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/scripts/train.py", line 100, in train_loop
    trainer.train()
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/engine/trainer.py", line 255, in train
    loss, loss_dict, metrics_dict = self.train_iteration(step)
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/utils/profiler.py", line 127, in inner
    out = func(*args, **kwargs)
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/engine/trainer.py", line 467, in train_iteration
    with torch.autocast(device_type=cpu_or_cuda_str, enabled=self.mixed_precision):
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/torch/amp/autocast_mode.py", line 201, in __init__
    raise RuntimeError('User specified autocast device_type must be \'cuda\' or \'cpu\'')
RuntimeError: User specified autocast device_type must be 'cuda' or 'cpu'
```

**Additional context**
I have a macbook pro M1 Max. To install nerfstudio I install functorch and pymeshlab and then `pip install nerfstudio`. It is not specified in the docs, but it worked after some trial and error. I needed to use python3.9, though. The data processing failed when not using the exhaustive matching method. 

The error seems to be a problem with the implementation of mixed precision. MPS does not support mixed precision, and so calls to autocast must be removed or casted to cpu. If instead of using device type MPS I chose CPU it does not work either. In that case it throws this error:

```
Traceback (most recent call last):
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/bin/ns-train", line 10, in <module>
    sys.exit(entrypoint())
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/scripts/train.py", line 261, in entrypoint
    main(
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/scripts/train.py", line 246, in main
    launch(
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/scripts/train.py", line 189, in launch
    main_func(local_rank=0, world_size=world_size, config=config)
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/scripts/train.py", line 100, in train_loop
    trainer.train()
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/engine/trainer.py", line 255, in train
    loss, loss_dict, metrics_dict = self.train_iteration(step)
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/utils/profiler.py", line 127, in inner
    out = func(*args, **kwargs)
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/engine/trainer.py", line 468, in train_iteration
    _, loss_dict, metrics_dict = self.pipeline.get_train_loss_dict(step=step)
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/utils/profiler.py", line 127, in inner
    out = func(*args, **kwargs)
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/pipelines/base_pipeline.py", line 281, in get_train_loss_dict
    model_outputs = self._model(ray_bundle)  # train distributed data parallel model if world_size > 1
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/models/base_model.py", line 142, in forward
    return self.get_outputs(ray_bundle)
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/models/nerfacto.py", line 277, in get_outputs
    ray_samples, weights_list, ray_samples_list = self.proposal_sampler(ray_bundle, density_fns=self.density_fns)
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/model_components/ray_samplers.py", line 50, in forward
    return self.generate_ray_samples(*args, **kwargs)
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/model_components/ray_samplers.py", line 603, in generate_ray_samples
    density = density_fns[i_level](ray_samples.frustums.get_positions())
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/fields/base_field.py", line 67, in density_fn
    density, _ = self.get_density(ray_samples)
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/fields/density_fields.py", line 105, in get_density
    self.mlp_base(positions_flat).view(*ray_samples.frustums.shape, -1).to(positions)
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/torch/nn/modules/container.py", line 217, in forward
    input = module(input)
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/field_components/mlp.py", line 178, in forward
    return self.pytorch_fwd(in_tensor)
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/nerfstudio/field_components/mlp.py", line 164, in pytorch_fwd
    for i, layer in enumerate(self.layers):
  File "/Users/joseperezcano/Desktop/Tenerife/nerf teide/colmap/build/.venv/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1614, in __getattr__
    raise AttributeError("'{}' object has no attribute '{}'".format(
AttributeError: 'MLP' object has no attribute 'layers'
```

**Comment by Jerry-Master** [comment]
The first bug can be fixed by adding the line `cpu_or_cuda_str = 'cpu' if cpu_or_cuda_str == 'mps' else cpu_or_cuda_str`. I am investigating how to solve the second.

**Comment by Jerry-Master** [comment]
Okay, so the second bug is solved in #2124 but apparently it has not reached the releases yet. I am going to check if the first bug is also solved, and if not I will create a pull request.
